content/courses/advanced-ada/chapters/access_types.rst

```diff
@@ -1577,17 +1577,732 @@ type) and display its information by calling the :ada:`Show` procedure.
+see how the accessibility rules use those levels.
+
+
+Operations on Access Types
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In very simple terms, the accessibility rules say that:
+
+- operations on access types at the same accessibility level are legal;
+
+- assigning or converting to a deeper level is legal;
+
+Otherwise, operations targeting objects at a *less-deep* level are illegal.
+
+For example, :ada:`L0_IA := new Integer'(22)` and :ada:`L1_IA := L1_Var'Access`
+are legal because we're operating at the same accessibility level. Also,
+:ada:`L1_IA := L0_Var'Access` is legal because :ada:`L1_IA` is at a deeper
+level than :ada:`L0_Var'Access`.
+
+However, many operations in the code example are illegal. For instance,
+:ada:`L0_IA := L1_Var'Access` and :ada:`L0_IA_2 := L1_Var'Access` are illegal
+because the target objects in the assignment are *less* deep.
+
+Note that the :ada:`L0_IA := L1_IA` assignment is mainly illegal because the
+access types don't match. (Of course, in addition to that, assigning
+:ada:`L1_Var'Access` to :ada:`L0_IA` is also illegal in terms of accessibility
+rules.)
+
+
+Conversion between Access Types
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The same rules apply to the conversion between access types. In the
+code example, the :ada:`L0_Integer_Access (L1_IA)` conversion is illegal
+because the resulting object is less deep. That being said, conversions on the
+same level are fine:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Same_Level_Conversion
+
+    procedure Show_Same_Level_Conversion is
+       type L1_Integer_Access is access all Integer;
+       type L1_B_Integer_Access is access all Integer;
+
+       L1_IA   : L1_Integer_Access;
+       L1_B_IA : L1_B_Integer_Access;
+
+       L1_Var  : aliased Integer;
+    begin
+       L1_IA := L1_Var'Access;
+
+       L1_B_IA := L1_B_Integer_Access (L1_IA);
+       --         ^^^^^^^^^^^^^^^^^^^
+       --         LEGAL: conversion from
+       --                L1 access object to
+       --                L1 access object
+    end Show_Same_Level_Conversion;
 
+Here, we're converting from the :ada:`L1_Integer_Access` type to the
+:ada:`L1_B_Integer_Access`, which are both at the same level.
+
+
+.. _Adv_Ada_Dangling_References:
+
+Dangling References
+~~~~~~~~~~~~~~~~~~~
+
+An access value that points to a non-existent object is called a dangling
+reference. Later on, we'll discuss how dangling references may occur using
+:ref:`unchecked deallocation <Adv_Ada_Unchecked_Deallocation_Dangling_References>`.
+
+Dangling references are created when we have an access value pointing to an
+object whose lifetime has ended, so it becomes a  non-existent object. This
+could occur, for example, when an access value still points to an object
+:ada:`X` that has gone out of scope.
+
+As mentioned in the previous section, the accessibility rules of the Ada
+language prevent that such situations ever happen! In fact, whenever possible,
```

Comment:
The phrase "language prevent that such situations ever happen!" is a little awkward in English. Maybe something like this:  "language ensure that such situations never happen!"

Reply:
Thanks! As a non-native speaker, I'm always grateful for this kind of corrections. :-)